core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -859,6 +847,29 @@ private[spark] class ExecutorAllocationManager(
       allocationManager.synchronized {
         // Clear unschedulableTaskSets since atleast one task becomes schedulable now
         unschedulableTaskSets.remove(stageAttempt)
+        removeStageFromResourceProfileIfUnused(stageAttempt)
+      }
+    }
+
+    def removeStageFromResourceProfileIfUnused(stageAttempt: StageAttempt): Unit = {
+      if (!stageAttemptToNumRunningTask.contains(stageAttempt) &&
+          !stageAttemptToNumTasks.contains(stageAttempt) &&
+          !stageAttemptToNumSpeculativeTasks.contains(stageAttempt) &&
+          !stageAttemptToTaskIndices.contains(stageAttempt) &&
+          !stageAttemptToSpeculativeTaskIndices.contains(stageAttempt) &&
+          !unschedulableTaskSets.contains(stageAttempt)
```

Comment:
we don't need to check for unschedulableTaskSets. if a attempt gets into there it should have stageAttemptToNumTasks . If its been stage completed then stageAttemptToNumTasks would be removed but we shouldn't still be trying to schedule it.